File: dsa_leetcode_beginner/0049-group-anagrams/0049-group-anagrams.cpp
// C++ implementation of group anagrams
#include <iostream>
#include <vector>
#include <unordered_map>
#include <string>
using namespace std;

vector<vector<string>> groupAnagrams(vector<string>& strs) {
    unordered_map<string, vector<string>> anagramMap;
    
    for (string word : strs) {
        vector<int> count(26, 0);
        for (char c : word) {
            count[c - 'a']++;
        }
        string key;
        for (int i = 0; i < 26; ++i) {
            key += string(1, count[i] + '0');
        }
        anagramMap[key].push_back(word);
    }
    
    vector<vector<string>> result;
    for (auto& pair : anagramMap) {
        result.push_back(pair.second);
    }
    return result;
}

int main() {
    vector<string> strs = {"eat", "tea", "tan", "ate", "nat", "bat"};
    vector<vector<string>> result = groupAnagrams(strs);
    for (const auto& group : result) {
        for (const auto& word : group) {
            cout << word << " ";
        }
        cout << endl;
    }
    return 0;
}

File: dsa_leetcode_beginner/0049-group-anagrams/0049-group-anagrams.js
// JavaScript implementation of group anagrams
function groupAnagrams(strs) {
    let anagramMap = new Map();
    
    strs.forEach(word => {
        let count = new Array(26).fill(0);
        for (let i = 0; i < word.length; i++) {
            count[word.charCodeAt(i) - 'a'.charCodeAt(0)]++;
        }
        let key = count.join('#');
        if (!anagramMap.has(key)) {
            anagramMap.set(key, []);
        }
        anagramMap.get(key).push(word);
    });
    
    return Array.from(anagramMap.values());
}

const strs = ["eat", "tea", "tan", "ate", "nat", "bat"];
console.log(groupAnagrams(strs));
